taskweaver/ces/environment.py

```diff
@@ -94,92 +98,253 @@ def pre_start_kernel(
+            if "JUPYTER_CONTROL_PORT" in env:
+                km.control_port = int(env["JUPYTER_CONTROL_PORT"])
+            if "JUPYTER_HB_PORT" in env:
+                km.hb_port = int(env["JUPYTER_HB_PORT"])
+            if "JUPYTER_IOPUB_PORT" in env:
+                km.iopub_port = int(env["JUPYTER_IOPUB_PORT"])
+            if "JUPYTER_KERNEL_IP" in env:
+                km.ip = env["JUPYTER_KERNEL_IP"]
         return km, kernel_name, kernel_id
 
 
+class EnvMode(enum.Enum):
+    SubProcess = "subprocess"
+    InsideContainer = "inside_container"
+    OutsideContainer = "outside_container"
+
+
 class Environment:
     def __init__(
         self,
         env_id: Optional[str] = None,
         env_dir: Optional[str] = None,
+        env_mode: Optional[EnvMode] = EnvMode.SubProcess,
+        port_start_inside_container: Optional[int] = 12345,
     ) -> None:
         self.session_dict: Dict[str, EnvSession] = {}
         self.id = get_id(prefix="env") if env_id is None else env_id
         self.env_dir = env_dir if env_dir is not None else os.getcwd()
+        self.mode = env_mode
+        if self.mode == EnvMode.SubProcess or self.mode == EnvMode.InsideContainer:
+            self.multi_kernel_manager = TaskWeaverMultiKernelManager(
+                default_kernel_name="taskweaver",
+                kernel_spec_manager=KernelSpecProvider(),
+            )
+            if self.mode == EnvMode.InsideContainer:
+                file_handler = logging.FileHandler("env.log")
+                file_handler.setLevel(logging.DEBUG)
+                file_handler.setFormatter(formatter)
+                logger.addHandler(file_handler)
+
+        elif self.mode == EnvMode.OutsideContainer:
+            try:
+                import docker
+                import docker.errors
+            except ImportError:
+                raise ImportError("docker package is required for container-based kernel.")
 
-        self.multi_kernel_manager = self.init_kernel_manager()
+            try:
+                self.docker_client = docker.from_env()
+            except docker.errors.DockerException as e:
+                raise docker.errors.DockerException(f"Failed to connect to Docker daemon: {e}. ")
+
+            self.session_container_dict: Dict[str, str] = {}
+            self.port_start_inside_container = port_start_inside_container
+        else:
+            raise ValueError(f"Unsupported environment mode {env_mode}")
+        atexit.register(self.clean_up)
+        logger.info(f"Environment {self.id} is created.")
 
     def clean_up(self) -> None:
+        logger.info(f"Environment {self.id} is cleaning up.")
         for session in self.session_dict.values():
             try:
                 self.stop_session(session.session_id)
             except Exception as e:
                 logger.error(e)
 
-    def init_kernel_manager(self) -> MultiKernelManager:
-        atexit.register(self.clean_up)
```

Comment:
do we need to somehow register env clean up somewhere as well?

Reply:
I did some experiments and it seems working fine.